src/plugins/data/public/search/fetch/extract_warnings.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import { estypes } from '@elastic/elasticsearch';
+import { i18n } from '@kbn/i18n';
+import { SearchResponseWarning } from '../types';
+
+/**
+ * @internal
+ */
+export function extractWarnings(
+  rawResponse: estypes.SearchResponse | undefined
+): SearchResponseWarning[] {
+  if (!rawResponse) {
+    return [];
+  }
+
+  const warnings: SearchResponseWarning[] = [];
+
+  let timedOut: SearchResponseWarning | undefined;
+  if (rawResponse.timed_out === true) {
+    timedOut = {
+      type: 'timed_out',
+      isTimeout: true,
+      message: i18n.translate('data.search.searchSource.fetch.requestTimedOutNotificationMessage', {
+        defaultMessage: 'Data might be incomplete because your request timed out',
+      }),
+    };
+  }
+  if (timedOut) {
+    warnings.push(timedOut);
+  }
+
+  if (rawResponse._shards && rawResponse._shards.failed) {
+    const isShardFailure = true;
+    const message = i18n.translate(
+      'data.search.searchSource.fetch.shardsFailedNotificationMessage',
+      {
+        defaultMessage: '{shardsFailed} of {shardsTotal} shards failed',
+        values: {
+          shardsFailed: rawResponse._shards.failed,
+          shardsTotal: rawResponse._shards.total,
+        },
+      }
+    );
+    const text = i18n.translate(
+      'data.search.searchSource.fetch.shardsFailedNotificationDescription',
+      { defaultMessage: 'The data you are seeing might be incomplete or wrong.' }
+    );
+
```

Comment:
push the generic shard failirue into the warnings array after all the other warnings

Reply:
The generic shard failure warning is a fallback, when `_shards.failed` > 0, but there are no reasons provided under `_shards.failures`. I moved this code to inspector RequestAdapter and added comments. Hope it is now clear.

Reply:
with multiple shard failure reasons this will show multiple notifications, which is different from previous behaviour.